PR Multi Fidelity Tabular Benchmarks

PhMueller opened the pull request.

Neeratyoy committed:
Adding sample RF space for tabular collection design

Neeratyoy committed:
Adding sample RF space for tabular collection design

Neeratyoy committed:
Placeholder SVM benchmark to interface tabular data collection

Neeratyoy committed:
Placeholder SVM benchmark to interface tabular data collection

Neeratyoy committed:
Writing common ML benchmark class for tabular collection

Neeratyoy committed:
Writing common ML benchmark class for tabular collection

Neeratyoy committed:
Adding placeholder for HistGradientBoostedClassifier

Neeratyoy committed:
Adding placeholder for HistGradientBoostedClassifier

Neeratyoy committed:
Minor code cleaning

Neeratyoy committed:
Minor code cleaning

Neeratyoy committed:
Reformatting output dict + option to add more metrics

Neeratyoy committed:
Reformatting output dict + option to add more metrics

Neeratyoy committed:
Removing redundant import

Neeratyoy committed:
Removing redundant import

Neeratyoy committed:
Decoupling storage of costs for each metric

Neeratyoy committed:
Decoupling storage of costs for each metric

Neeratyoy committed:
Including test scores in objective

Neeratyoy committed:
Including test scores in objective

Neeratyoy committed:
Documenting the structure of information in each fn eval.

Neeratyoy committed:
Documenting the structure of information in each fn eval.

Neeratyoy committed:
Some decisions on lower bound for subsample fidelity

Neeratyoy committed:
Some decisions on lower bound for subsample fidelity

Neeratyoy committed:
AbstractBenchmark update for fidelity option + including XGBoost

Neeratyoy committed:
AbstractBenchmark update for fidelity option + including XGBoost

Neeratyoy committed:
Merge branch 'thesis-paper' of https://github.com/Neeratyoy/HPOBench into thesis-paper

Neeratyoy committed:
Option to load data splits from disk

Neeratyoy committed:
Reordering data load to work for different cases

Neeratyoy committed:
Updating source of SVM HP range

Neeratyoy committed:
Adding Tabular Benchmark class

Neeratyoy committed:
Adding TabularBenchmark interface + easy import

Neeratyoy committed:
Adding LR space

Neeratyoy committed:
Standardizing fidelity space definitions

Neeratyoy committed:
Standardizing HPs + Adding NN space

Neeratyoy committed:
Small placeholder for testing

Neeratyoy committed:
Updating NN HP space + Helper function for TabularBenchmark

Neeratyoy committed:
Adding fidelity range retrieval utility to TabularBenchmark

Neeratyoy committed:
Enforcing subsample lower bound check inside objective

Neeratyoy committed:
Bug fix + adding precicion as metric

Neeratyoy committed:
Fixing param spaces and model building for LR, SVM

Neeratyoy committed:
TabularBenchmark edit to read compressed files and query a dataframe

Neeratyoy committed:
Not evaluating training set to save time

Neeratyoy committed:
Fidelity change for trees + NN space change

Neeratyoy committed:
Final RF space

Neeratyoy committed:
Final XGB space

Neeratyoy committed:
Final HistGB space

Neeratyoy committed:
Finalizing RF, XGB, NN

Neeratyoy committed:
TabularBenchmark edit to process only table and metadata

PhMueller committed:
Merge remote-tracking branch 'origin/development' into PR_Multi-fidelity-tabular-benchmarks

PhMueller committed:
Merge remote-tracking branch 'origin/development' into PR_Multi-fidelity-tabular-benchmarks

PhMueller committed:
TabularBenchmark

- Rearrange the benchmark.
- Move data parts to a data manager. It is able to download the data from the web, if it is not present on the local machine.
- Enforce the API structure

PhMueller committed:
Pycodestyle

PhMueller committed:
Flake8

PhMueller committed:
Adapt ML Benchmark Template to fit with current API

PhMueller committed:
Corret Datamanager.

But how to download the task data to disk?

PhMueller committed:
Finalize HistGB Benchmarks

- split them into multiple benchmarks according to their fidelity spaces.

PhMueller committed:
Write OpenML Datamanager

PhMueller committed:
Unify interface for the other ml benchmarks.

PhMueller committed:
Flake + Pep

PhMueller committed:
Add Container Interface

PhMueller committed:
Mark `task_id` as required.

PhMueller committed:
Adapt Interfaces

PhMueller committed:
Fix minor errors.

PhMueller committed:
Fix minor errors.

PhMueller committed:
Pylint

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/entry_point.py:
```diff
@@ -0,0 +1,24 @@
+from hpobench.benchmarks.ml_mmfb.histgb_benchmark import HistGBSearchSpace0Benchmark, HistGBSearchSpace1Benchmark, \
```
Do we need an extra .py file here? 
Doing something like `this` should make all the available classes accessible under `hpolib.benchmarks.ml` or `hpolib.benchmarks.ml_mmfb`? 
So a `from hpolib.benchmarks.ml import *` should suffice when using the benchmark right?

PhMueller replied:
Normally yes. But I need this for the container interface.

Neeratyoy replied:
Can we then also have `__init__` populated as normal?

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/histgb_benchmark.py:
```diff
@@ -0,0 +1,136 @@
+            CS.UniformIntegerHyperparameter(
+                'max_depth', lower=6, upper=30, default_value=6, log=True
+            ),
+            # TODO: The parameter max_leaf_node is not accepted. Changed it from max_leaf_node to max_leaf_nodes
```
You're right, you can remove the TODO here.

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/histgb_benchmark.py:
```diff
@@ -0,0 +1,136 @@
+            CS.UniformIntegerHyperparameter(
+                'max_leaf_nodes', lower=2, upper=64, default_value=32, log=True
+            ),
+            # TODO: The parameter eta is not accepted. Do you mean learning_rate? Changed it from eta to learning_rate
```
Same.

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/histgb_benchmark.py:
```diff
@@ -0,0 +1,136 @@
+        subsample = fidelity2[subsample_choice]
+        return ntrees, subsample
+
+    def init_model(self, config: Dict, fidelity: Dict = None, rng: Union[int, np.random.RandomState, None] = None):
```
Why the `Dict` enforcing of config and fidelity? 
I would rather have it be flexible and take both ConfigSpace and Dict if Dict needs to be supported explicitly

PhMueller replied:
Alright. I'll change it. However, the wrapper in the objective function converts the `CS.Configuration` to `Dict` in the last step. So basically, I have changed it after it crashed. It was called with a Dict and then the Dict does not have a .get_dictionary() function. 

Is then something like a check in the beginning of the function fine:
``` 
if isinstance(config, CS.Configuration):
    config = config.get_dictionary()
```

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/histgb_benchmark.py:
```diff
@@ -0,0 +1,136 @@
+        return fidelity_space
+
+
+__all__ = [HistGBSearchSpace0Benchmark, HistGBSearchSpace1Benchmark,
```
The design to bypass the static function definition is nice. I too have some local commits pending rn but it's a slightly different design where I am overloading the function with a different parameter space. One reason I like your design is what I was actually aiming for since we then have just one definition for the parameter ranges, which is always safer and robust to future changes. I think I can safely then forego my local commits wrt this but I'd like to have a bit more structure and interpretability with the Search Space overloading as follows:

* HistGBBenchmark:  `ntrees_choice='variable`, `subsample_choice='variable'` (multi-multi-fidelity version by default)
* HistGBBenchmarkBB:  `ntrees_choice='fixed`, `subsample_choice='fixed'` (black-box version)
* HistGBBenchmarkMF:  `ntrees_choice='variable`, `subsample_choice='fixed'` (multi-fidelity version)

I think this design can hold us in good stead in the long run too. Having a second Mult-fidelity version with `subsample='variable'` and `ntrees='fixed'` is not quite needed I feel. Especially once we collect data for multi-multi-fidelity, all other cases will be subsumed by it. 

Also, you could then incorporate the `variable, variable` case into the HistGBBaseBenchmark and just name it `HistGBBenchmark`. I'm thinking of having the default class be the actual multi-multi-fidelity benchmark case. With the black-box and multi-fidelity cases, inherited from it by overloading *only* the fidelity space.

This design is what I'll recommend for the other model spaces too!

PhMueller replied:
Sounds good. I will change it.

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/lr_benchmark.py:
```diff
@@ -0,0 +1,127 @@
+
+
+__all__ = [LRSearchSpace0Benchmark, LRSearchSpace1Benchmark,
+           LRSearchSpace2Benchmark, LRSearchSpace3Benchmark]
```
should become,
```python
__all__ = [LRBenchmark, LRBenchmarkMF, LRBenchmarkBB]
```

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/nn_benchmark.py:
```diff
@@ -0,0 +1,141 @@
+
+
+__all__ = [NNSearchSpace0Benchmark, NNSearchSpace1Benchmark,
+           NNSearchSpace2Benchmark, NNSearchSpace3Benchmark]
```
should become,
```python
__all__ = [NNBenchmark, NNBenchmarkMF, NNBenchmarkBB]
```

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/rf_benchmark.py:
```diff
@@ -0,0 +1,137 @@
+
+
+__all__ = [RandomForestSearchSpace0Benchmark, RandomForestSearchSpace1Benchmark,
+           RandomForestSearchSpace2Benchmark, RandomForestSearchSpace3Benchmark]
```
should become,
```python
__all__ = [RandomForestBenchmark, RandomForestBenchmarkMF, RandomForestBenchmarkBB]
```

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/svm_benchmark.py:
```diff
@@ -0,0 +1,94 @@
+        fidelity_space = CS.ConfigurationSpace(seed=seed)
+        fidelity_space.add_hyperparameter(
+            # parameterize the fraction of data to subsample
+            SVMBaseBenchmark._get_fidelity_choices(subsample_choice='fixed')
```
this is also `fixed` the same as the above class
one of them should be `variable`

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/svm_benchmark.py:
```diff
@@ -0,0 +1,94 @@
+from hpobench.dependencies.ml.ml_benchmark_template import MLBenchmark
+
+
+class SVMBaseBenchmark(MLBenchmark):
```
For SVM, I'd create three classes again, to keep the parity of definitions
where `SVMBenchmark` and `SVMBenchmarkMF` are the same definitions (since only 1 fidelity)
while `SVMBenchmarkBB` is the black-box case

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/tabular_benchmark.py:
```diff
@@ -0,0 +1,230 @@
+from hpobench.util.data_manager import TabularDataManager
+
+
+class BaseTabularBenchmark(AbstractBenchmark):
```
Do we really need a Base Class here?
I'll add more reasons why I think this is not required in the comment below.

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/tabular_benchmark.py:
```diff
@@ -0,0 +1,230 @@
+        return cs
+
+
+class OriginalTabularBenchmark(BaseTabularBenchmark):
```
I don't think we need this class definition. I'd rather have the utilities of this class as a function along the lines of `get_original_configuration/fidelity_space` inside `TabularBenchmark`. 
Given that the `TabularBenchmark` is designed to query the tabular data using the discretized configspace, it doesn't make sense to instantiate an entire object of a different class with the `original` space since the `objective_function` call won't work with this `original` non-discretized space. 
Exposing `get_original_configuration/fidelity_space` is just for the user to see the original space that was used to derive the discrete space returned by the current definition inside `TabularBenchmark`. 

As a result, I think we can do away with this class, and combine the BaseClass and the TabularBenchmark into just one class.

Let me know if you still need this design for the sake of containerization and the need to inherit `AbstractBenchmark`.

PhMueller replied:
Ah, get it now! Thanks for the clarification. 
I have removed the second class. (original) and moved the original config space/fidelity space to the meta-information as a workaround. Maybe, we could open an issue and expose a new function in the container interface that enables us to query the function later.

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/xgboost_benchmark.py:
```diff
@@ -0,0 +1,151 @@
+
+
+__all__ = [XGBoostSearchSpace0Benchmark, XGBoostSearchSpace1Benchmark,
+           XGBoostSearchSpace2Benchmark, XGBoostSearchSpace3Benchmark]
```
Same as above: `XGBoostBenchmark`, `XGBoostBenchmarkBB`, `XGBoostBenchmarkMF`

Neeratyoy commented on hpobench/dependencies/ml/data_manager.py:
```diff
@@ -0,0 +1,199 @@
+from hpobench import config_file
+
+
+class OpenMLDataManager(DataManager):
```
I am not testing this...trust you with this interface ;) 
but looks good, maybe later while building the README, we can include an example of its use?

and about the location, given we might expand the benchmark and using OpenML preprocessing might extend beyond `ml`, so how about placing it one level above?

Neeratyoy commented on hpobench/dependencies/ml/ml_benchmark_template.py:
```diff
@@ -0,0 +1,290 @@
+import time
```
my main question is about the location of this file
especially if we are creating a subfolder `ml` under `dependencies`, might as well have it in the benchmark folder?

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/lr_benchmark.py:
```diff
@@ -0,0 +1,127 @@
+                 data_path: Union[str, None] = None):
+
+        super(LRBaseBenchmark, self).__init__(task_id, rng, valid_size, data_path)
+        self.cache_size = 500  # TODO: Do we need this?
```
I was recommended to have it since SVM can become super slow for larger datasets
is it causing any issues with containerization?

PhMueller replied:
No, but I just did not found where it is used. :-)

Neeratyoy replied:
https://github.com/automl/HPOBench/blob/PR_Multi-fidelity-tabular-benchmarks/hpobench/benchmarks/ml_mmfb/svm_benchmark.py#L70

PhMueller replied:
Jep. but this is the lr - benchmark. And it does not inherit from the svm benchmark.

Neeratyoy replied:
oh then I totally got it wrong
`cache_size` is apparently applicable only for SVM

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/rf_benchmark.py:
```diff
@@ -0,0 +1,137 @@
+        assert subsample_choice in ['fixed', 'variable']
+
+        fidelity1 = dict(
+            fixed=CS.Constant('n_estimators', value=100),  # TODO: is the default value here 100 or 512?
```
`512`
fixed all of these silly things in my local commit :P

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/tabular_benchmark.py:
```diff
@@ -0,0 +1,230 @@
+    def _search_dataframe(self, row_dict, df):
+        query_stmt = self._build_query(row_dict)
+        result = df.query(query_stmt)
+        # TODO: What happens in this case? The objective function raises a TypeError.
```
I am not sure about this scenario and when it arises actually

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/tabular_benchmark.py:
```diff
@@ -0,0 +1,230 @@
+            return None
+        return result.iloc[0].loc['result']
+
+        # TODO: This created an out-of-bounds error. The idx mask should have been 2d, but was 1d.
```
I will try and do some local testing, and some benchmarking of speed
it is surprising that it fails
it would be great if you can help me point to the failing case for me to debug

PhMueller replied:
I will change this back to the masking via `.values`

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/tabular_benchmark.py:
```diff
@@ -0,0 +1,230 @@
+        cost_key = f"{evaluation}_scores"
+
+        key_path = dict()
+        # TODO: Dicts are unordered. This does not have to have an effect.
```
not sure what you meant here, i.e., the effect of `np.sort`?

PhMueller replied:
I just thought that we can remove the ` np.sort(..)`. 
Because if you go in a increasing order through the hp-names and then insert them into a dictionary that has no ordering, this operation does not change anything. 
Or why are you using the sort function here?

Neeratyoy replied:
hmm not 100% sure since there are a couple of reasons
for a dict, yes I agree, doesn't make sense, or rather is immaterial
however, I think the DataFrame columns are in this order and when I am looking up HP names and creating the query order, it's important to be in that order 
`np.sort` probably is not required since I think the `get_hyperparameter_names` does so internally or that's how dict keys are returned probably
however, since a lot of steps in the pipeline (benchmark collection) depend on it, having `np.sort` was kind of a strict check I had enforced myself to ensure parity

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/xgboost_benchmark.py:
```diff
@@ -0,0 +1,151 @@
+        assert subsample_choice in ['fixed', 'variable']
+
+        fidelity1 = dict(
+            fixed=CS.Constant('n_estimators', value=100),  # TODO: Should this be 1000 or 100?
```
sorry again, `1000`

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/xgboost_benchmark.py:
```diff
@@ -0,0 +1,151 @@
+        if isinstance(fidelity, CS.Configuration):
+            fidelity = fidelity.get_dictionary()
+
+        # TODO: This seems to be wrong. (AND-condition)
```
yeahup clearly wrong :/ glad this doesn't affect the benchmarks collected so far :p

Neeratyoy reviewed (commented):
I have left some comments, the main change required I think is the class name definitions.

PhMueller committed:
Init Model can handle now Configurations

Neeratyoy commented on hpobench/benchmarks/ml_mmfb/tabular_benchmark.py:
```diff
@@ -0,0 +1,230 @@
+    def _seeds_used(self) -> List:
+        return self.table.seed.unique().tolist()
+
+    def sample_hyperparamer(self, n: int = 1) -> Union[CS.Configuration, List]:
```
a typo from my end, could you please fix this 
should ofc be `sample_hyperparameter`

PhMueller committed:
PR Requests: Rename Classes

PhMueller committed:
PR Requests: Move dependencies to correct directory

PhMueller committed:
PR Requests: Tabular Benchmarks - Remove unnecessary class definition

PhMueller committed:
PR Requests: Minor improvments

PhMueller committed:
PR Requests: Update upper bounds of the fidelities

PhMueller committed:
PR Requests: Remove OriginalTabBenchmarks

PhMueller committed:
PR Requests: Revert the query function

PhMueller committed:
PR Requests: Minor improvements

PhMueller committed:
Pycodestyle

PhMueller committed:
Add missing requirements

PhMueller committed:
Minor Improvements

- cast return values to float
- improve the __all__ vars

PhMueller committed:
ADD container recipes

PhMueller committed:
PR: Fix path in tabular data loader

PhMueller committed:
PR: Remove casting configspace to np.floats

PhMueller committed:
PR: Move everything back from ml_mmfb/ to ml/

PhMueller committed:
PR: Remove pybnn from the init.

This would cause an error since importing stuff for the tab benchmarks would require the pybnn stuff.

PhMueller committed:
PR: Cleanup

PhMueller committed:
PR: Fix Tests

PhMueller committed:
PR Fix URLs and dependencies

PhMueller committed:
PR Fix URLs and dependencies

PhMueller committed:
PR Fix URLs and dependencies

PhMueller committed:
PR Fix tests

PhMueller committed:
New Urls

PhMueller committed:
Trigger Rebuild.

PhMueller committed:
Fix Dataloader Assertion
